src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -0,0 +1,176 @@
+        private const byte Prefix_Request = 7;
+        private const byte Prefix_IdList = 6;
+
+        private const long OracleRequestPrice = 0_50000000;
+
+        public override int Id => -4;
+        public override string Name => "Oracle";
+
+        internal OracleContract()
+        {
+            Manifest.Features = ContractFeatures.HasStorage;
+        }
+
+        [ContractMethod(0, CallFlags.AllowModifyStates)]
+        private void Finish(ApplicationEngine engine)
+        {
+            Transaction tx = (Transaction)engine.ScriptContainer;
+            OracleResponse response = tx.Attributes.OfType<OracleResponse>().First();
+            OracleRequest request = GetRequest(engine.Snapshot, response.Id);
+            StackItem userData = BinarySerializer.Deserialize(request.UserData, engine.MaxStackSize, engine.MaxItemSize, engine.ReferenceCounter);
+            engine.CallFromNativeContract(null, request.CallbackContract, request.CallbackMethod, request.Url, userData, response.Success, response.Result);
+        }
+
+        [ContractMethod(0_01000000, CallFlags.AllowStates)]
+        public ECPoint[] GetOracleNodes(StoreView snapshot)
+        {
+            return snapshot.Storages[CreateStorageKey(Prefix_NodeList)].GetInteroperable<NodeList>().ToArray();
+        }
+
+        private UInt256 GetOriginalTxid(ApplicationEngine engine)
+        {
+            Transaction tx = (Transaction)engine.ScriptContainer;
+            OracleResponse response = tx.Attributes.OfType<OracleResponse>().FirstOrDefault();
+            if (response is null) return tx.Hash;
+            OracleRequest request = GetRequest(engine.Snapshot, response.Id);
+            return request.OriginalTxid;
+        }
+
+        public OracleRequest GetRequest(StoreView snapshot, ulong id)
+        {
+            return snapshot.Storages.TryGet(CreateStorageKey(Prefix_Request).Add(id))?.GetInteroperable<OracleRequest>();
+        }
+
+        public IEnumerable<OracleRequest> GetRequests(StoreView snapshot)
+        {
+            return snapshot.Storages.Find(new byte[] { Prefix_Request }).Select(p => p.Value.GetInteroperable<OracleRequest>());
+        }
+
+        public IEnumerable<OracleRequest> GetRequestsByUrl(StoreView snapshot, string url)
+        {
+            IdList list = snapshot.Storages.TryGet(CreateStorageKey(Prefix_IdList).Add(GetUrlHash(url)))?.GetInteroperable<IdList>();
+            if (list is null) yield break;
+            foreach (ulong id in list)
+                yield return snapshot.Storages[CreateStorageKey(Prefix_Request).Add(id)].GetInteroperable<OracleRequest>();
+        }
+
+        private static byte[] GetUrlHash(string url)
+        {
+            return Crypto.Hash160(Utility.StrictUTF8.GetBytes(url));
+        }
+
+        internal override void Initialize(ApplicationEngine engine)
+        {
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_NodeList), new StorageItem(new NodeList()));
```

Comment:
Init by default with the validator values?

Reply:
The IP addresses of consensus nodes should not be exposed. It is dangerous for the consensus nodes to be oracle nodes. I think it's no problem that there is no oracle node in initial state.

Reply:
The problem it's for testing, this could delay every test because you need to set the oracle nodes first, maybe we can create a json file that include the initial TX in the genesis block?

Reply:
Or add default oracle nodes in protocol.json?

Reply:
It's good for me too

Reply:
Can this be done in UT? Maybe we should keep empty for the mainnet.